File: 백준/Gold/1937. 욕심쟁이 판다/욕심쟁이 판다.py
import sys

sys.setrecursionlimit(10 ** 6)
input = sys.stdin.readline

N = int(input())

arr = [list(map(int, input().split())) for _ in range(N)]
di = [0, 0, 1, -1]
dj = [1, -1, 0, 0]
dp = [[0] * N for _ in range(N)]


def check(y, x):
    if dp[y][x] != 0:
        return dp[y][x]

    dp[y][x] = 1 
    for d in range(4):
        ny, nx = y + di[d], x + dj[d]
        if 0 <= ny < N and 0 <= nx < N:
            if arr[y][x] < arr[ny][nx]:
                dp[y][x] = max(dp[y][x], check(ny, nx) + 1)

    return dp[y][x]


ans = 0
for i in range(N):
    for j in range(N):
        ans = max(ans, check(i, j))

print(ans)

File: 백준/Gold/1937. 욕심쟁이 판다/욕심쟁이 판다.cc
#include <iostream>
#include <algorithm>
using namespace std;

int n;
int arr[500][500];
int dp[500][500];

int dy[4] = { -1,0,1,0 };
int dx[4] = { 0,1,0,-1 };

int dfs(int y, int x) {
    if (dp[y][x] != 0) return dp[y][x];

    dp[y][x] = 1;

    for (int d = 0; d < 4; d++) {
        int ny = y + dy[d];
        int nx = x + dx[d];
        if (ny < 0 || nx < 0 || ny >= n || nx >= n) continue;
        if (arr[ny][nx] <= arr[y][x]) continue; 

        dp[y][x] = max(dp[y][x], dfs(ny, nx) + 1);
    }

    return dp[y][x];
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    cin >> n;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            cin >> arr[i][j];

    int ans = 0;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            ans = max(ans, dfs(i, j));

    cout << ans;
    return 0;
}
